contracts/utils/cryptography/EIP712.sol

```diff
@@ -24,17 +25,21 @@ import "./ECDSA.sol";
  *
  * _Available since v3.4._
  */
-abstract contract EIP712 {
+abstract contract EIP712 is ERC5267 {
     /* solhint-disable var-name-mixedcase */
     // Cache the domain separator as an immutable value, but also store the chain id that it corresponds to, in order to
     // invalidate the cached domain separator if the chain id changes.
     bytes32 private immutable _CACHED_DOMAIN_SEPARATOR;
     uint256 private immutable _CACHED_CHAIN_ID;
     address private immutable _CACHED_THIS;
 
+    string private _NAME;
+    string private _VERSION;
```

Comment:
It's a little frustrating that we have to put this in storage when it's going to be very short strings that could be immutable. There are things we could do to have an immutable and a storage fallback. Are we open to introducing that complexity here?

Ideally we would define a library or structure that would do the optimization and fallback mechanism but I don't know if it's possible.

Reply:
I'm thinking something like this (rough sketch):

```solidity
ShortString immutable _name;
ShortStringFallback _nameFallback; // struct { string value; }
...
_name = _setShortWithFallback(name, _nameFallback);
...
string memory name = _getShortWithFallback(_name, _nameFallback);
```

Reply:
I'm open to it, but it it really worth it.

- Is setting an immutable dynamically really less expensive than writing to storage once ?
- This should only be read "off-chain" by the eip712Domain accessor. There is no real "sload" to avoid.

Also, on upgradeable contract this will be storage anyway and the "short" strings will be optimized into a single slot.

Reply:
>Also, on upgradeable contract this will be storage anyway and the "short" strings will be optimized into a single slot.

That's still better though! Normal strings require at least two slots.



> Is setting an immutable dynamically really less expensive than writing to storage once ?

Pretty sure it is significantly less expensive but we would need to check.



> This should only be read "off-chain" by the eip712Domain accessor. There is no real "sload" to avoid.

Yeah totally. The concern is just the SSTOREs during deployment. 2 additional string variables, i.e. 4 SSTORE, would result in 40k gas. Looking at [these benchmarks](https://github.com/alephao/solidity-benchmarks/blob/main/benchmarks/0.8.17/ERC20.md#deployment) it seems like this would be an additional 10% cost in deployment for the Permit variant.

Reply:
We can see it in the gas report in this PR. This is the increase to deployment cost:
![image](https://user-images.githubusercontent.com/481465/213300014-27616a5a-e237-4f0b-abef-e5c844f83342.png)

Reply:
> That's still better though! Normal strings require at least two slots.

My understanding is that string of length <=31 used a single slot.

Reply:
I can confirm that small strings only use one storage slot:

- see [this contract ](https://goerli.etherscan.io/address/0x18c6b374f8cba732fb36e4b03f10a2320584aea9#code)
  - the `push` function write a string into a mapping, this is the only sstore
- see [this transaction](https://goerli.etherscan.io/tx/0xfbe70e846da08e0efa3b3e7771b43a015bafd85a9f8b4505d207466625cd3df3)
  - only one slot is modified: the string fits into a single slot

![Capture d’écran du 2023-01-19 11-17-39](https://user-images.githubusercontent.com/2432299/213416581-c20dadaa-1625-432f-af34-2ce02a9fe373.png)

Reply:
so its "just" 2 sstore in almost all cases

Reply:
Nice, I did not know that! It's in the [Solidity docs](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html#bytes-and-string) too.

So comparing with the benchmarks repo it would be "just" 5% increase. It's still the case that we see a 14% increase in our gas report here.